### PR TITLE
feat(sql): add allowlist sql validator

### LINKED ISSUE #1334: Build the allowlist-primary SQL validator
## Summary

Build the `sqlglot`-based validator that enforces the read-only contract **before** any execution. It is **allowlist-primary**: only a curated set of safe scalar/aggregate functions is permitted; everything unknown is rejected by default. A blocklist remains as defense-in-depth.

## Context

- `sqlglot` (ClickHouse dialect) is a new backend dependency (pin `>=25,<27`; pin an exact minor in production and re-run the suite before upgrading).
- Tenant scoping (`project_id`) is supplied server-side; the validator never sees it.

## Implementation

`validate(sql) -> exp.Query` enforces:
- parseable; non-empty; **single statement** (filter `None`, count remainder); top-level is `exp.Query` (SELECT/UNION) — reject all DML/DDL/`exp.Command` (SYSTEM/SET/OPTIMIZE/USE/KILL/…).
- **Table whitelist:** every `exp.Table` is a bare identifier in `{spans, traces}` or a CTE alias; reject table functions (non-`Identifier` `this`), db/catalog-qualified names, internal `*_public_v1` names, and **`FINAL`** (`table.args["final"]`). Skip `ARRAY JOIN` right-hand sides (column nodes, not tables).
- **CTE-shadow:** reject a CTE named `spans`/`traces`.
- **`project_id` blocked** as a column (incl. via table alias) and as an output alias (`… AS project_id`).
- **`SETTINGS` blocked** (top-level `tree.args["settings"]` and nested `exp.Settings`).
- **Function gate:** `ALLOWED_FUNCTIONS` frozenset (aggregates, date/time, type-conversion, conditional, math, string, array, window, safe JSON/hash). For each `exp.Anonymous` (`node.name`) and `exp.Func` subclass (`node.sql_name()`), the lowercased name must be in the allowlist; unknown → reject. `exp.Cast`/`exp.TryCast`/`exp.Extract`/`exp.Lambda` are skipped (non-callable nodes). Operators, `CAST`, `INTERVAL`, `CASE`, `IN`, `BETWEEN`, `LIKE` are non-`Func` nodes and allowed. **Blocklist** (`BLOCKED_FUNCTIONS` + `dictget*`/`joinget*` prefixes) wins over the allowlist.

## Test plan

`backend/tests/services/sql/test_validator.py` (parametrized):
- **Allowed:** aggregates, `quantile`, math, date bucketing, type conversion, conditional/`CASE`, string, array, `CAST`/operators, window functions with allowed aggregates, lambdas, JSON/hash, CTEs, subqueries, `UNION ALL`.
- **Rejected (one each):** DML/DDL/SYSTEM/SET; multi-statement + trailing-`;` injection; unknown/qualified/system/internal-view tables; `FINAL`; table functions; **unknown functions** (allowlist enforcement — ≥3 cases); blocked scalars (`sleep`/`getSetting`/`currentUser`/…); `dictGet*`/`joinGet*`; `project_id` column + alias; CTE named `spans`/`traces`; `SETTINGS` (top-level + nested).
- Window with unknown aggregate → reject; lambda hiding `sleep` → reject; `ARRAY JOIN` column not misidentified as a table.

## Acceptance criteria

- [ ] `ALLOWED_FUNCTIONS` (frozenset, lowercased canonical names), `BLOCKED_FUNCTIONS`, `BLOCKED_PREFIXES` exported; blocklist wins over allowlist.
- [ ] Every `ALLOWED` query validates; every `REJECTED` query raises `SqlValidationError` with a sanitized, non-empty message (no view names / project_id / raw SQL).
- [ ] **Unknown functions are rejected by default** (allowlist primary), incl. inside window functions and lambda bodies.
- [ ] `FINAL`, nested `SETTINGS`, `project_id` alias, and `ARRAY JOIN` gaps are covered.
- [ ] `sqlglot` added to deps and locked; the trailing-`;` behavior is empirically verified and documented.
- [ ] No DB/network access; `test_validator.py` passes.

### PR DESCRIPTION
**Pull request by dark-sorceror**
## Summary

Fixes #1334 

Adds the first enforcement layer of the SQL query/export feature: a `sqlglot`-based validator that checks user SQL against a read-only, allowlist-primary contract **before** anything is executed. Only a small, curated set of safe analytical functions is permitted; everything unknown is rejected by default, with a blocklist as defense-in-depth that always wins over the allowlist.

It builds on the public SQL schema contract — it reads the curated table list as the single source of truth for which tables a query may touch, so the validator and the exposed schema can never drift apart.

## What it enforces

`validate(sql)` rejects anything that isn't a single read-only SELECT/UNION over the curated `spans` / `traces` tables:

- **Read-only only** — rejects INSERT/UPDATE/DELETE/ALTER/CREATE/DROP/TRUNCATE/OPTIMIZE/SYSTEM/SET and any non-query command; a single statement only. A bare trailing `;` is a terminator and parses as one statement, so it is accepted and discarded at parse time; anything that follows it is a second statement and is rejected.
- **Table whitelist** — only the bare logical tables `spans` and `traces`. Rejects unknown tables, database-qualified names (e.g. `system.tables`, `default.spans`), internal curated-view names, table functions (`url`/`s3`/`remote`/`numbers`/…), and the `FINAL` modifier. Common table expressions are allowed, but a CTE named `spans`/`traces` (a shadowing attack) is rejected.
- **Tenant safety** — the tenant identifier is rejected as a column, qualified column, and output alias. Scoping is always applied server-side, never from user SQL.
- **No escape hatches** — `SETTINGS` (top-level and nested) and output redirects (`FORMAT`, `INTO OUTFILE`) are rejected.
- **Function gate (allowlist-primary)** — every function call must be in the conservative allowlist or it's rejected, enforced across the whole query including window-function and lambda bodies. A blocklist of dangerous families (filesystem/network/remote access, other-database connectors, cluster, denial-of-service, and information-disclosure functions) wins over the allowlist.

Error messages are sanitized: a validation error never echoes the raw SQL, a tenant identifier, or internal view names — including the parse-error path.

## What's included

- `backend/rest/services/sql/validator.py` — `validate()` plus the exported `ALLOWED_FUNCTIONS`, `BLOCKED_FUNCTIONS`, and `BLOCKED_PREFIXES` constants
- `backend/rest/services/sql/errors.py` — the `SqlValidationError` type
- `sqlglot` added to dependencies and locked

The validator is pure: no database, network, configuration, or file I/O access.

## Notes

- The allowlist is intentionally conservative (50 functions today). It can be widened further as real query patterns are observed.
- **Map access for the curated `metadata` column** — the schema contract surfaces `metadata` as a Map, so `metadata['user_id']` (a subscript, no function involved) already passed. `mapKeys` / `mapValues` / `mapContains` are allowlisted alongside it, because a Map you can key but not inspect is only usable by someone who already knows the keys. The wider array surface that composes with `mapKeys()` — `arrayJoin` and friends — stays rejected, asserted by a test rather than left to inference.
- Project-scoped query rewriting, read-only database execution, the HTTP endpoint, the database views, and the CLI are **not** part of this change.

## Test plan

`tests/services/sql/test_sql_validator.py` — parametrized allowed/rejected matrices plus targeted edge cases:

- [x] Allowed: `count`/`sum`/`quantile` aggregates, `now() - INTERVAL`, CTEs, subqueries, `UNION ALL`, `CASE`/`CAST`/arithmetic/`IN`/`BETWEEN`/`LIKE`
- [x] Rejected: every DML/DDL/command form; multi-statement, including anything following a trailing `;`; unknown/qualified/system/internal-view tables; table functions; `FINAL`; three or more unknown functions; blocked scalars (`sleep`/`getSetting`/`currentUser`); dictionary/join-get prefixes; tenant column in projection/filter/alias; CTE shadow; `SETTINGS`
- [x] Error messages leak no raw SQL, tenant identifier, or view names (including a parse-error regression test)
- [x] Existing schema tests still pass

121 tests pass (105 validator + 16 schema), and the full backend suite is green on this branch. Lint and format clean.

---
## Summary by cubic
Adds a `sqlglot`-based SQL validator that enforces the read-only, allowlist-approved SELECT/UNION contract over the public `spans`/`traces` tables before anything executes. Fixes #1334.

- Reads the public table list from the shared schema contract so the validator and exposed schema can't drift apart.
- Resolves CTE references scope-aware and case-insensitively so a same-named CTE can't smuggle a real table past the allowlist.
- Blocks `project_id` as a column, qualifier, and alias in every position, including CTE aliases, derived-table aliases, and aliases on CTE references, plus `scope_*` reserved bound parameters; error messages never leak raw SQL, the tenant column, or internal view names.
- Rejects `{name:Identifier}`-typed query parameters and, until parameter binding is wired, refuses all other bound parameters too.
- Function gate is allowlist-first with a blocklist override that always wins; covers aggregates, date bucketing, time-range, string, window functions, and `metadata` map accessors, with `count(*)`, `EXISTS`, and boolean `AND`/`OR` handling resolving correctly.
- Rejects qualified names, table functions, `FINAL`, `SETTINGS`, `FORMAT`, and any statement after a trailing `;`.

**Dependencies**
- Added `sqlglot`, pinned to exactly 26.33.0 because the allowlist depends on its function-name normalization.

<sup>Written for commit 13b42326618f8841cb92ae6b4999f3b6f820f2c8. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/traceroot-ai/traceroot/pull/1354?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: backend/rest/services/sql/validator.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+                raise SqlValidationError(
+                    "Qualified table names (db.table / catalog.db.table) are not allowed"
+                )
+
+            # Internal view names must not be queried directly.
+            if table_name in _INTERNAL_VIEWS:
+                raise SqlValidationError("Internal view names may not be queried directly")
+
+            # CTE alias references are allowed unconditionally.
+            if table_name in cte_names:
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py]: Resolved in 703b7ad

Good catch. The validator collected CTE names into one flat global set, so a CTE named `evil` in any scope made an out-of-scope physical table `evil` pass the `{spans, traces}` whitelist. Empirically reproduced: `SELECT * FROM evil` was correctly rejected, but the same reference wrapped with a sibling-scope `WITH evil AS (...)` was wrongly accepted — letting an attacker smuggle any bare table name past Layer 1.

I replaced the flat global set with sqlglot's `build_scope` resolver — a table reference is only treated as a CTE when a CTE of that name is visible in its own scope (`scope.cte_sources`). Fails closed if scope resolution errors.

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

File: backend/rest/services/sql/validator.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            if id(node) in cte_ref_ids:
+                continue
+
+            # Must be a known public table.
+            if table_name not in public_table_names:
+                raise SqlValidationError("Table is not in the allowed public schema")
+
+        # 7. project_id as a column reference (SELECT project_id / WHERE project_id = …).
+        if isinstance(node, exp.Column) and node.name.lower() == "project_id":
+            raise SqlValidationError("Column 'project_id' is not accessible")
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py]: Resolved in 30126d8

Genericized both `project_id` column/alias errors to "Access to a restricted column is not allowed" so the message no longer echoes the reserved column name; added a regression test.

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

File: backend/rest/services/sql/validator.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    for scope in root_scope.traverse():
+        for table in scope.tables:
+            if table.name.lower() in scope.cte_sources:
+                cte_ref_ids.add(id(table))
+
+    # Walk the full AST (including subqueries, CTEs, window bodies) for all
+    # remaining policy checks.
+    for node in tree.walk():
+        # 8. SETTINGS at any nesting level.
+        if isinstance(node, exp.Select) and node.args.get("settings"):
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py]: Resolved in 8d53d93

The check only inspected `exp.Select`, missing modifiers on `Union/Subquery` roots. Fix: check `exp.Query` (base of `Select/Union/Subquery`) for both SETTINGS and FORMAT. TDD: added the two bypass cases as failing tests first, then fixed; both now rejected, normal UNION/subquery still allowed, no regression.

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

File: backend/rest/services/sql/validator.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+        raise SqlValidationError("SQL could not be analyzed for table access") from exc
+
+    if root_scope is None:
+        raise SqlValidationError("SQL could not be analyzed for table access")
+
+    # Mark every exp.Table node whose name is visible as a CTE in its own scope.
+    cte_ref_ids: set[int] = set()
+    for scope in root_scope.traverse():
+        for table in scope.tables:
+            if table.name.lower() in scope.cte_sources:
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py]: Resolved in b770218

Lowercase both sides of the CTE-visibility comparison (consistent with every other identifier check in the validator).

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

File: backend/rest/services/sql/validator.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+            # Must be a known public table.
+            if table_name not in public_table_names:
+                raise SqlValidationError("Table is not in the allowed public schema")
+
+        # 7. project_id as a column reference (SELECT project_id / WHERE project_id = …)
+        #    or as an output alias (… AS project_id). Both use the same generic
+        #    wording so the error never names the reserved tenant-scoping column
+        #    (keeps messages sanitized / avoids confirming internal schema).
+        if isinstance(node, exp.Column) and node.name.lower() == "project_id":
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py]: Resolved in 2792c30

`project_id` is now rejected in every identifier position (column, column table-qualifier, output alias, table alias). Normal aliases, joins, and mixed-case CTEs still validate.

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

File: backend/rest/services/sql/validator.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        statements = [s for s in sqlglot.parse(sql, read="clickhouse") if s is not None]
+    except Exception as exc:
+        # Never interpolate the sqlglot exception: it embeds the offending SQL
+        # fragment (which can carry a project_id or other tenant data). Keep the
+        # original on the exception chain for server-side debugging only.
+        raise SqlValidationError("SQL statement could not be parsed") from exc
+
+    # 3. Exactly one statement required.
+    # (Trailing ';' → one statement; 'SELECT 1; SELECT 2' → two → rejected.)
+    if len(statements) != 1:
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py:226]: False positive.

1. No injection is possible. The single-statement gate counts non-None parsed nodes, so a trailing `;` (which yields a trailing empty node) still passes as one statement, but `; SELECT 1` and `; DROP TABLE` spans are both rejected (2 statements). A trailing semicolon cannot smuggle a second statement past the check.
2. The semicolon never reaches ClickHouse. The rewriter re-renders the executed SQL from the parsed AST, so `SELECT count() FROM spans;` becomes `SELECT count() FROM spans_public_v1(...)` — no semicolon in the output. It's discarded before execution.
3. It's intentional and spec-compliant. The spec's Layer-1 rule is "reject more than one statement (a trailing ; that produces a second node counts)" — i.e., a bare trailing ; is explicitly allowed. The code documents this, and it's the normal, user-friendly behavior (rejecting `SELECT …;` would be an annoying regression).

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

File: tests/services/sql/test_sql_validator.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+    for prefix in BLOCKED_PREFIXES:
+        assert isinstance(prefix, str)
+        assert prefix == prefix.lower(), f"BLOCKED_PREFIXES entry not lowercased: {prefix!r}"
+
+
+def test_blocklist_wins_over_allowlist() -> None:
+    # Manually add a blocked name into ALLOWED_FUNCTIONS would not help;
+    # here we verify the symbolic contract: blocked set and allowed set
+    # are both present as public constants.
+    assert BLOCKED_FUNCTIONS.isdisjoint(ALLOWED_FUNCTIONS), (
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply tests/services/sql/test_sql_validator.py:269]: Valid — fixed in d2d6e715d. The old test asserted the two sets are disjoint, which is a real invariant but leaves the precedence branch untested by construction: if the allowlist were ever checked first, nothing would fail.

Split into two tests. `test_blocked_and_allowed_sets_do_not_overlap` keeps the invariant; `test_blocklist_wins_when_a_name_reaches_both_sets` monkeypatches `ALLOWED_FUNCTIONS` to include `sleep` — forcing the overlap the invariant forbids — and asserts `validate("SELECT sleep(10) FROM spans")` still raises.

File: backend/rest/services/sql/validator.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py:239]: Working as designed, and now pinned by a test rather than left to inference.

A bare trailing `;` is a terminator, not a second statement: sqlglot yields one statement, the single-statement gate passes, and the rewriter renders the AST rather than the caller's text — so the semicolon never reaches ClickHouse. What the contract rejects is a second *statement*, which the rejection matrix already covers with `SELECT 1; SELECT 2` and `SELECT 1; DROP TABLE spans`; both raise.

Rejecting the bare terminator would only make well-formed client SQL fail for a character that is discarded at parse time. d2d6e715d adds `test_bare_trailing_semicolon_is_one_statement` documenting the boundary explicitly.

File: backend/rest/services/sql/validator.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+        #     key `project_id` or anything in the `scope_` namespace. Blocking
+        #     these here (Layer 1, case-insensitive) means a user placeholder can
+        #     never collide with or override the scope bind, independent of how
+        #     the service later merges the bind map.
+        if isinstance(node, exp.Placeholder):
+            var = node.this
+            param_name = var.name if isinstance(var, exp.Expression) else str(var or "")
+            lowered = param_name.lower()
+            if lowered == "project_id" or lowered.startswith(_RESERVED_PARAM_PREFIX):
+                raise SqlValidationError("Bound parameters may not use a reserved name")
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py:400]: Valid — fixed in 33c8b95b5. Confirmed both halves before fixing:

```
SELECT {col:Identifier} FROM spans          -> ACCEPTED
SELECT * FROM spans WHERE {c:Identifier}=1  -> ACCEPTED
SELECT * FROM {tbl:Identifier}              -> AttributeError: 'Var' object has no attribute 'lower'
```

The third is the worse half: an identifier placeholder in table position reached the table walk as a `Var` and escaped as a raw `AttributeError` rather than a sanitized `SqlValidationError` — a 500 with a stack trace instead of a rejected query.

The gate now runs before the table walk and keys on the AST shape: sqlglot models a value parameter's kind as an `exp.DataType` and the identifier form as the plain string `Identifier`, so anything that is not a `DataType` — including a bare `{name}` — is refused with "Query parameters must be typed as values", which echoes neither the parameter name nor the SQL. Callers' own value parameters (`{v:String}`, `{c:Float64}`) still pass, and the rewriter's injected `{scope_project_id:String}` is unaffected.

One correction to the impact as written: `project_id` is not in either curated view's projection, so `column=project_id` would have failed at execution with an unknown-identifier error rather than returning the tenant column. The table-position case was the real exposure — the rewriter never rewrites a placeholder, so the substituted name would have reached ClickHouse unscoped, backstopped only by the read-only user's grants.

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

File: backend/rest/services/sql/validator.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+        #    table alias (FROM spans AS project_id). All share the same generic
+        #    wording so the error never names the reserved column. (Tenant scoping
+        #    itself is enforced by the rewriter's view injection, independent of
+        #    these checks; this keeps the surface tidy and free of footguns.)
+        if isinstance(node, exp.Column) and (
+            node.name.lower() == "project_id" or node.table.lower() == "project_id"
+        ):
+            raise SqlValidationError("Access to a restricted column is not allowed")
+
+        if isinstance(node, exp.Alias) and node.alias.lower() == "project_id":
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py:377]: Valid — fixed in 4ec273b9e. Confirmed the gap first:

```
WITH project_id AS (SELECT 1 AS x) SELECT x FROM project_id        -> ACCEPTED
SELECT span_id FROM (SELECT span_id FROM spans) AS project_id      -> ACCEPTED
SELECT span_id FROM spans AS project_id                            -> rejected
SELECT span_id AS project_id FROM spans                            -> rejected
WITH project_id AS (...) SELECT project_id.x FROM project_id       -> rejected
```

The check covered the positions that *reference* a name (column, qualifier, output alias, table alias) but not the two that *introduce* one — `exp.CTE` and `exp.Subquery` aliases.

No leak in the gap itself: a CTE or derived table named `project_id` exposes no tenant column, the qualified form was already refused, and the rewriter never rewrites a CTE reference. What it did break is the module's own stated contract — the comment claims the name is blocked in EVERY identifier position, and an inconsistency like that is what a later change reasons from. Both alias positions now raise the same generic error that names no column, and two tests assert ordinary CTE and derived-table aliases still pass.

File: backend/rest/services/sql/validator.py [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+        #    table alias (FROM spans AS project_id), and the two alias positions
+        #    that introduce a NAME rather than reference one — a CTE
+        #    (WITH project_id AS …) and a derived table
+        #    (FROM (…) AS project_id). All share the same generic
+        #    wording so the error never names the reserved column. (Tenant scoping
+        #    itself is enforced by the rewriter's view injection, independent of
+        #    these checks; this keeps the surface tidy and free of footguns.)
+        if isinstance(node, exp.Column) and (
+            node.name.lower() == "project_id" or node.table.lower() == "project_id"
+        ):
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py:374]: Not reachable — and the conditional in the finding ("limited if the executed public views omit `project_id`") is exactly the design, not a hope.

`SELECT *` never resolves against a physical table. The rewriter retargets every public table reference before execution:

```
SELECT * FROM spans
  -> SELECT * FROM spans_public_v1(project_id = {scope_project_id: String}) AS spans
```

and the curated view projects an explicit column list that excludes `project_id`, the `ch_*` bookkeeping columns, the `input`/`output`/`metadata` blobs, and the `source`/`is_evaluation` control flags. Verified against the real view DDL on ClickHouse 25.2.2.39:

```
DESCRIBE (SELECT * FROM spans_public_v1(project_id = 'pA'))
span_id   String
trace_id  String
name      String
metadata  Map(LowCardinality(String), String)

SELECT project_id FROM spans_public_v1(project_id = 'pA')
Code: 47. DB::Exception: Unknown expression identifier `project_id` ... (UNKNOWN_IDENTIFIER)
```

The other tenant's row was absent from the same run. So the column policy in the validator is defence in depth against a *named* reference, not the control that keeps `*` safe — that is the curated view, which is issue (4) of this stack and pinned by its own projection test. This is the epic's stated guarantee: `SELECT *` yields the curated analytical column set, never the physical table's.

**Comment by dark-sorceror** [comment]
@cubic-dev-ai review

File: backend/rest/services/sql/validator.py [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+                )
+
+            # Internal view names must not be queried directly.
+            if table_name in _INTERNAL_VIEWS:
+                raise SqlValidationError("Internal view names may not be queried directly")
+
+            # CTE alias references are allowed only when the CTE is visible
+            # in this table's own scope (scope-aware; see cte_ref_ids above).
+            if id(node) in cte_ref_ids:
+                continue
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py:355]: Valid, fixed in 9f999ae9c. Confirmed the asymmetry first:

```
WITH c AS (SELECT 1 AS x) SELECT x FROM c AS project_id   -> ACCEPTED
SELECT span_id FROM spans AS project_id                   -> rejected
```

`continue` ends the walk for that node, so check 7 never ran for a table reference that resolved to a CTE: the same reserved alias was refused or allowed depending only on what it aliased. The check now runs before the early exit, with tests for both the refusal and for an ordinary alias on a CTE reference still passing.

This is the third position in that family to need closing, after the CTE and derived-table alias definitions. The pattern is that every early exit in the table walk is a place the reserved-name contract can leak out of.

File: backend/rest/services/sql/validator.py [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+    policy violation.  The message never echoes the raw SQL, a ``project_id``,
+    or internal view names.
+    """
+    # 1. Reject empty / whitespace-only.
+    if not sql or not sql.strip():
+        raise SqlValidationError("SQL statement must not be empty")
+
+    # 2. Parse as ClickHouse dialect; any parse error is a validation failure.
+    try:
+        statements = [s for s in sqlglot.parse(sql, read="clickhouse") if s is not None]
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply backend/rest/services/sql/validator.py:230]: Same verdict as the previous two times this was raised, but this round I have fixed what keeps generating it rather than only restating the answer.

The behaviour is deliberate and the code says so at both the module docstring and the statement-count check: a bare trailing `;` is a terminator, `sqlglot.parse` returns one statement, and the rewriter renders from the AST, so the character never reaches ClickHouse. `SELECT 1; SELECT 2` and `SELECT 1; DROP TABLE spans` are two statements and are rejected. `test_bare_trailing_semicolon_is_one_statement` pins the boundary.

What did not say so was this PR description, which read "multi-statement and trailing-`;` injection are rejected". That parses as a claim that a bare terminator is refused, which is the contract this finding keeps measuring the code against. Reworded to state the actual rule, so the description and the code now agree.

**Review by XinweiHe — APPROVED** [review]
